File: backend/app/src/main/kotlin/net/animeta/backend/service/admin/ImageService.kt
package net.animeta.backend.service.admin

import com.google.common.io.Files
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.Resource
import org.springframework.core.io.WritableResource
import org.springframework.http.HttpMethod
import org.springframework.http.client.OkHttp3ClientHttpRequestFactory
import org.springframework.stereotype.Service
import org.springframework.web.client.ResponseExtractor
import org.springframework.web.client.RestTemplate
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

@Service
class ImageService(
    @Value("\${animeta.media.root_location}") private val mediaRoot: Resource,
    private val annService: AnnService
) {
    private val restTemplate = RestTemplate(OkHttp3ClientHttpRequestFactory().apply {
        setConnectTimeout(0)
        setReadTimeout(0)
        setWriteTimeout(0)
    })

    fun downloadAnnPoster(annId: String, outFile: File) {
        val anime = annService.getMetadata(annId)!!
        val images = anime.select("info[@type=\"Picture\"] img")
        var fullsrc: String? = null
        for (image in images) {
            val src = image.attr("src")
            if (src.contains("full")) {
                fullsrc = src
            } else if (src.contains("max") && fullsrc == null) {
                fullsrc = src
            }
        }
        if (fullsrc == null) {
            throw IllegalStateException("full image url not found")
        }
        download(fullsrc.replaceFirst("http://", "https://"), outFile)
    }

    fun generateThumbnail(file: File, thumbFile: File, removeAnnWatermark: Boolean = false) {
        var w = 233
        var h = 318
        val annWatermarkHeight = 13

        // Retina
        w *= 2
        h *= 2

        val args = mutableListOf<String>("convert", file.absolutePath)
        if (removeAnnWatermark) {
            args.add("-crop")
            args.add("+0-$annWatermarkHeight")
        }
        args.add("-resize")
        args.add("${w}x$h^")
        args.add("-gravity")
        args.add("north")
        args.add("-crop")
        args.add("${w}x$h+0+0")
        args.add(thumbFile.absolutePath)

        if (ProcessBuilder(args).inheritIO().start().waitFor() != 0) {
            throw Exception("resize failed")
        }

        try {
            ProcessBuilder(listOf("jpegoptim", "--max", "40", "--totals", thumbFile.absolutePath))
                    .inheritIO().start().waitFor()
        } catch (e: IOException) {
            // ignore
        }
    }

    fun download(url: String, dest: File) {
        restTemplate.execute(url, HttpMethod.GET, null,
                ResponseExtractor {
                    it.use {
                        Files.asByteSink(dest).writeFrom(it.body)
                    }
                })
    }

    fun upload(source: File, path: String) {
        try {
            FileOutputStream(File(mediaRoot.file, path))
        } catch (e: UnsupportedOperationException) {
            (mediaRoot.createRelative(path) as WritableResource).outputStream
        }.use {
            Files.asByteSource(source).copyTo(it)
        }
    }
}

File: backend-ts/src/services/admin/image.service.ts
import { Injectable } from "@nestjs/common";
import got from 'got'
import * as fs from 'fs'
import { pipeline } from 'stream';
import { promisify } from "util";
import { spawn } from "child_process";
import { Storage, Bucket } from '@google-cloud/storage';
import { ConfigService } from "@nestjs/config";
import { getAnnMetadata } from "./ann_metadata_cache";
import cheerio from 'cheerio';

const pipelinePromise = promisify(pipeline)

function normalizeBasePath(basePath: string): string {
  if (basePath[0] === '/')
    basePath = basePath.substring(1)

  if (basePath === '')
    return ''
  
  if (basePath[basePath.length - 1] !== '/')
    return basePath + '/'
  else
    return basePath
}

@Injectable()
export class ImageService {
  private storage: Storage
  private mediaStorageBucket: Bucket
  private mediaStorageBasePath: string

  constructor(configService: ConfigService) {
    this.storage = new Storage()
    const mediaStorageUrl = new URL(configService.get('ANIMETA_MEDIA_STORAGE_URL')!)
    if (mediaStorageUrl.protocol !== 'gs:')
      throw new Error(`Unsupported storage: ${mediaStorageUrl.protocol}`)
    this.mediaStorageBucket = this.storage.bucket(mediaStorageUrl.hostname)
    this.mediaStorageBasePath = normalizeBasePath(mediaStorageUrl.pathname)
  }

  async downloadAnnPoster(annId: string, outFile: string) {
    const anime = await getAnnMetadata(annId)
    const images = anime.find('info[type="Picture"] img')
    let fullsrc: string | null = null
    for (const image of images.toArray()) {
      let src = cheerio(image).attr('src')
      if (!src) continue
      if (src.includes('full'))
        fullsrc = src
      else if (src.includes('max') && !fullsrc)
        fullsrc = src
    }
    if (!fullsrc)
      throw Error('full image url not found')
    await this.download(fullsrc.replace('http://', 'https://'), outFile)
  }

  async download(url: string, dest: string) {
    await pipelinePromise(
      got.get(url, {isStream: true}),
      fs.createWriteStream(dest)
    )
  }

  async generateThumbnail(file: string, thumbFile: string, removeAnnWatermark: boolean = false) {
    let w = 233
    let h = 318
    const annWatermarkHeight = 13

    // Retina
    w *= 2
    h *= 2

    const args = [file]
    if (removeAnnWatermark) {
      args.push('-crop', `+0-${annWatermarkHeight}`)
    }
    args.push("-resize", `${w}x${h}^`)
    args.push("-gravity", "north")
    args.push("-crop", `${w}x${h}+0+0`)
    args.push(thumbFile)

    const exitCode = await new Promise((resolve, reject) => {
      const proc = spawn('convert', args, {stdio: 'inherit'})
      proc.on('error', err => reject(err))
      proc.on('close', code => resolve(code))
    })
    if (exitCode !== 0)
      throw new Error('resize failed')
    
    try {
      await new Promise((resolve, reject) => {
        const proc = spawn('jpegoptim', ["--max", "40", "--totals", thumbFile], {stdio: 'inherit'})
        proc.on('error', err => reject(err))
        proc.on('close', code => resolve(code))
      })
    } catch (e) {
      // ignore
    }
  }

  async upload(source: string, path: string) {
    await this.mediaStorageBucket
      .upload(source, {destination: this.mediaStorageBasePath + path})
  }
}
